src/main/scala/is/hail/utils/UInt.scala

```diff
@@ -46,25 +50,57 @@ class UInt private(val i: Int) extends AnyVal {
     if (l == r) 0 else if (l > r) 1 else -1
   }
 
-  def +(right: UInt): UInt = UInt(toLong + right.toLong)
-
-  def -(right: UInt): UInt = UInt(toLong - right.toLong)
-
-  def *(right: UInt): UInt = UInt(toLong * right.toLong)
-
-  def /(right: UInt): UInt = UInt(toLong / right.toLong)
-
-  def <=(right: UInt): Boolean = toLong <= right.toLong
-
-  def >=(right: UInt): Boolean = toLong >= right.toLong
-
-  def <(right: UInt): Boolean = toLong < right.toLong
-
-  def >(right: UInt): Boolean = toLong > right.toLong
-
-  def ==(right: UInt): Boolean = toLong == right.toLong
-
-  def !=(right: UInt): Boolean = toLong != right.toLong
+  def intRep: Int = i
+
+  def +(right: UInt, dummy: Boolean = false): UInt = UInt(toLong + right.toLong)
+  def +(right: Int): UInt = UInt(toLong + right)
+  def +(right: Double): Double = toDouble + right
+  def +(right: Long): Long = toLong + right
+
+  def -(right: UInt, dummy: Boolean = false): UInt = UInt(toLong - right.toLong)
+  def -(right: Int): UInt = UInt(toLong - right)
+  def -(right: Double): Double = toDouble - right
+  def -(right: Long): Long = toLong - right
+
+  def *(right: UInt, dummy: Boolean = false): UInt = UInt(toLong * right.toLong)
+  def *(right: Int): UInt = UInt(toLong * right)
+  def *(right: Double): Double = toDouble * right
+  def *(right: Long): Long = toLong * right
+
+  def /(right: UInt, dummy: Boolean = false): UInt = UInt(toLong / right.toLong)
+  def /(right: Int): UInt = UInt(toLong / right)
+  def /(right: Double): Double = toDouble / right
+  def /(right: Long): Long = toLong / right
+
+  def <=(right: UInt, dummy: Boolean = false): Boolean = toLong <= right.toLong
+  def <=(right: Int): Boolean = toLong <= right
+  def <=(right: Double): Boolean = toDouble <= right
+  def <=(right: Long): Boolean = toLong <= right
+
+  def >=(right: UInt, dummy: Boolean = false): Boolean = toLong >= right.toLong
+  def >=(right: Int): Boolean = toLong >= right
+  def >=(right: Double): Boolean = toDouble >= right
+  def >=(right: Long): Boolean = toLong >= right
+
+  def <(right: UInt, dummy: Boolean = false): Boolean = toLong < right.toLong
+  def <(right: Int): Boolean = toLong < right
+  def <(right: Double): Boolean = toDouble < right
+  def <(right: Long): Boolean = toLong < right
+
+  def >(right: UInt, dummy: Boolean = false): Boolean = toLong > right.toLong
+  def >(right: Int): Boolean = toLong > right
+  def >(right: Double): Boolean = toDouble > right
+  def >(right: Long): Boolean = toLong > right
+
+  def ==(right: UInt, dummy: Boolean = false): Boolean = toLong == right.toLong
+  def ==(right: Int): Boolean = toLong == right
+  def ==(right: Double): Boolean = toDouble == right
+  def ==(right: Long): Boolean = toLong == right
+
+  def !=(right: UInt, dummy: Boolean = false): Boolean = toLong != right.toLong
+  def !=(right: Int): Boolean = toLong != right
+  def !=(right: Double): Boolean = toDouble != right
+  def !=(right: Long): Boolean = toLong != right
```

Comment:
I think these are all correct.  If two integers are different sizes, they should be converted to the larger.  If one is unsigned and one is signed, they should be converted to unsigned, and integers should be converted to floating point.